File: leetcode/0135-candy.ts
/**
 * Distributes minimum number of candies to children based on ratings
 * Each child must receive at least one candy
 * Children with higher ratings get more candies than their neighbors
 * 
 * @param ratings - Array of children's ratings
 * @returns Minimum number of candies needed
 */
function candy(ratings: number[]): number {
    const n: number = ratings.length;
  
    // Track minimum candies needed based on left neighbor comparison
    const leftToRight: number[] = new Array(n).fill(1);
  
    // Track minimum candies needed based on right neighbor comparison
    const rightToLeft: number[] = new Array(n).fill(1);
  
    // First pass: left to right
    // If current child has higher rating than left neighbor, give one more candy
    for (let i: number = 1; i < n; i++) {
        if (ratings[i] > ratings[i - 1]) {
            leftToRight[i] = leftToRight[i - 1] + 1;
        }
    }
  
    // Second pass: right to left
    // If current child has higher rating than right neighbor, give one more candy
    for (let i: number = n - 2; i >= 0; i--) {
        if (ratings[i] > ratings[i + 1]) {
            rightToLeft[i] = rightToLeft[i + 1] + 1;
        }
    }
  
    // Calculate total candies needed
    // Take maximum of both passes to satisfy both left and right neighbor constraints
    let totalCandies: number = 0;
    for (let i: number = 0; i < n; i++) {
        totalCandies += Math.max(leftToRight[i], rightToLeft[i]);
    }
  
    return totalCandies;
}

File: leetcode/0135-candy.java
class Solution {
    public int candy(int[] ratings) {
        int n = ratings.length;
      
        // Array to store minimum candies needed considering left neighbor constraint
        int[] leftToRight = new int[n];
        // Array to store minimum candies needed considering right neighbor constraint
        int[] rightToLeft = new int[n];
      
        // Initialize both arrays with 1 (minimum candy per child)
        Arrays.fill(leftToRight, 1);
        Arrays.fill(rightToLeft, 1);
      
        // Left to right pass: ensure higher rated child gets more candy than left neighbor
        for (int i = 1; i < n; i++) {
            if (ratings[i] > ratings[i - 1]) {
                leftToRight[i] = leftToRight[i - 1] + 1;
            }
        }
      
        // Right to left pass: ensure higher rated child gets more candy than right neighbor
        for (int i = n - 2; i >= 0; i--) {
            if (ratings[i] > ratings[i + 1]) {
                rightToLeft[i] = rightToLeft[i + 1] + 1;
            }
        }
      
        // Calculate total candies needed
        int totalCandies = 0;
        for (int i = 0; i < n; i++) {
            // Take maximum to satisfy both left and right neighbor constraints
            totalCandies += Math.max(leftToRight[i], rightToLeft[i]);
        }
      
        return totalCandies;
    }
}